File: lib/codegen/meta-python/base/formats.py
"""
The cranelift.formats defines all instruction formats.

Every instruction format has a corresponding `InstructionData` variant in the
Rust representation of Cranelift IR, so all instruction formats must be defined
in this module.
"""
from __future__ import absolute_import
from cdsl.formats import InstructionFormat
from cdsl.operands import VALUE, VARIABLE_ARGS
from .immediates import imm64, uimm8, uimm32, ieee32, ieee64, offset32
from .immediates import boolean, intcc, floatcc, memflags, regunit, trapcode
from . import entities
from .entities import ebb, sig_ref, func_ref, stack_slot, heap, table

Unary = InstructionFormat(VALUE)
UnaryImm = InstructionFormat(imm64)
UnaryIeee32 = InstructionFormat(ieee32)
UnaryIeee64 = InstructionFormat(ieee64)
UnaryBool = InstructionFormat(boolean)
UnaryGlobalValue = InstructionFormat(entities.global_value)

Binary = InstructionFormat(VALUE, VALUE)
BinaryImm = InstructionFormat(VALUE, imm64)

# The select instructions are controlled by the second VALUE operand.
# The first VALUE operand is the controlling flag which has a derived type.
# The fma instruction has the same constraint on all inputs.
Ternary = InstructionFormat(VALUE, VALUE, VALUE, typevar_operand=1)

# Catch-all for instructions with many outputs and inputs and no immediate
# operands.
MultiAry = InstructionFormat(VARIABLE_ARGS)

NullAry = InstructionFormat()

InsertLane = InstructionFormat(VALUE, ('lane', uimm8), VALUE)
ExtractLane = InstructionFormat(VALUE, ('lane', uimm8))

IntCompare = InstructionFormat(intcc, VALUE, VALUE)
IntCompareImm = InstructionFormat(intcc, VALUE, imm64)
IntCond = InstructionFormat(intcc, VALUE)
FloatCompare = InstructionFormat(floatcc, VALUE, VALUE)
FloatCond = InstructionFormat(floatcc, VALUE)

IntSelect = InstructionFormat(intcc, VALUE, VALUE, VALUE)

Jump = InstructionFormat(ebb, VARIABLE_ARGS)
Branch = InstructionFormat(VALUE, ebb, VARIABLE_ARGS)
BranchInt = InstructionFormat(intcc, VALUE, ebb, VARIABLE_ARGS)
BranchFloat = InstructionFormat(floatcc, VALUE, ebb, VARIABLE_ARGS)
BranchIcmp = InstructionFormat(intcc, VALUE, VALUE, ebb, VARIABLE_ARGS)
BranchTable = InstructionFormat(VALUE, ebb, entities.jump_table)
BranchTableEntry = InstructionFormat(VALUE, VALUE, uimm8, entities.jump_table)
BranchTableBase = InstructionFormat(entities.jump_table)
IndirectJump = InstructionFormat(VALUE, entities.jump_table)

Call = InstructionFormat(func_ref, VARIABLE_ARGS)
CallIndirect = InstructionFormat(sig_ref, VALUE, VARIABLE_ARGS)
FuncAddr = InstructionFormat(func_ref)

Load = InstructionFormat(memflags, VALUE, offset32)
LoadComplex = InstructionFormat(memflags, VARIABLE_ARGS, offset32)
Store = InstructionFormat(memflags, VALUE, VALUE, offset32)
StoreComplex = InstructionFormat(memflags, VALUE, VARIABLE_ARGS, offset32)

StackLoad = InstructionFormat(stack_slot, offset32)
StackStore = InstructionFormat(VALUE, stack_slot, offset32)

# Accessing a WebAssembly heap.
HeapAddr = InstructionFormat(heap, VALUE, uimm32)

# Accessing a WebAssembly table.
TableAddr = InstructionFormat(table, VALUE, offset32)

RegMove = InstructionFormat(VALUE, ('src', regunit), ('dst', regunit))
CopySpecial = InstructionFormat(('src', regunit), ('dst', regunit))
RegSpill = InstructionFormat(
        VALUE, ('src', regunit), ('dst', entities.stack_slot))
RegFill = InstructionFormat(
        VALUE, ('src', entities.stack_slot), ('dst', regunit))

Trap = InstructionFormat(trapcode)
CondTrap = InstructionFormat(VALUE, trapcode)
IntCondTrap = InstructionFormat(intcc, VALUE, trapcode)
FloatCondTrap = InstructionFormat(floatcc, VALUE, trapcode)

# Finally extract the names of global values in this module.
InstructionFormat.extract_names(globals())

File: cranelift-codegen/meta/src/shared/formats.rs
use crate::cdsl::formats::{FormatRegistry, InstructionFormatBuilder as Builder};
use crate::shared::immediates::Immediates;
use crate::shared::OperandKinds;

pub fn define(imm: &Immediates, entities: &OperandKinds) -> FormatRegistry {
    // Shorthands for entities.
    let global_value = entities.by_name("global_value");
    let ebb = entities.by_name("ebb");
    let jump_table = entities.by_name("jump_table");
    let func_ref = entities.by_name("func_ref");
    let sig_ref = entities.by_name("sig_ref");
    let stack_slot = entities.by_name("stack_slot");
    let heap = entities.by_name("heap");
    let table = entities.by_name("table");

    let mut registry = FormatRegistry::new();

    registry.insert(Builder::new("Unary").value());
    registry.insert(Builder::new("UnaryImm").imm(&imm.imm64));
    registry.insert(Builder::new("UnaryImm128").imm(&imm.uimm128));
    registry.insert(Builder::new("UnaryIeee32").imm(&imm.ieee32));
    registry.insert(Builder::new("UnaryIeee64").imm(&imm.ieee64));
    registry.insert(Builder::new("UnaryBool").imm(&imm.boolean));
    registry.insert(Builder::new("UnaryGlobalValue").imm(global_value));

    registry.insert(Builder::new("Binary").value().value());
    registry.insert(Builder::new("BinaryImm").value().imm(&imm.imm64));

    // The select instructions are controlled by the second VALUE operand.
    // The first VALUE operand is the controlling flag which has a derived type.
    // The fma instruction has the same constraint on all inputs.
    registry.insert(
        Builder::new("Ternary")
            .value()
            .value()
            .value()
            .typevar_operand(1),
    );

    // Catch-all for instructions with many outputs and inputs and no immediate
    // operands.
    registry.insert(Builder::new("MultiAry").varargs());

    registry.insert(Builder::new("NullAry"));

    registry.insert(
        Builder::new("InsertLane")
            .value()
            .imm_with_name("lane", &imm.uimm8)
            .value(),
    );
    registry.insert(
        Builder::new("ExtractLane")
            .value()
            .imm_with_name("lane", &imm.uimm8),
    );

    registry.insert(Builder::new("IntCompare").imm(&imm.intcc).value().value());
    registry.insert(
        Builder::new("IntCompareImm")
            .imm(&imm.intcc)
            .value()
            .imm(&imm.imm64),
    );
    registry.insert(Builder::new("IntCond").imm(&imm.intcc).value());

    registry.insert(
        Builder::new("FloatCompare")
            .imm(&imm.floatcc)
            .value()
            .value(),
    );
    registry.insert(Builder::new("FloatCond").imm(&imm.floatcc).value());;

    registry.insert(
        Builder::new("IntSelect")
            .imm(&imm.intcc)
            .value()
            .value()
            .value(),
    );

    registry.insert(Builder::new("Jump").imm(ebb).varargs());
    registry.insert(Builder::new("Branch").value().imm(ebb).varargs());
    registry.insert(
        Builder::new("BranchInt")
            .imm(&imm.intcc)
            .value()
            .imm(ebb)
            .varargs(),
    );
    registry.insert(
        Builder::new("BranchFloat")
            .imm(&imm.floatcc)
            .value()
            .imm(ebb)
            .varargs(),
    );
    registry.insert(
        Builder::new("BranchIcmp")
            .imm(&imm.intcc)
            .value()
            .value()
            .imm(ebb)
            .varargs(),
    );
    registry.insert(Builder::new("BranchTable").value().imm(ebb).imm(jump_table));
    registry.insert(
        Builder::new("BranchTableEntry")
            .value()
            .value()
            .imm(&imm.uimm8)
            .imm(jump_table),
    );
    registry.insert(Builder::new("BranchTableBase").imm(jump_table));
    registry.insert(Builder::new("IndirectJump").value().imm(jump_table));

    registry.insert(Builder::new("Call").imm(func_ref).varargs());
    registry.insert(Builder::new("CallIndirect").imm(sig_ref).value().varargs());
    registry.insert(Builder::new("FuncAddr").imm(func_ref));

    registry.insert(
        Builder::new("Load")
            .imm(&imm.memflags)
            .value()
            .imm(&imm.offset32),
    );
    registry.insert(
        Builder::new("LoadComplex")
            .imm(&imm.memflags)
            .varargs()
            .imm(&imm.offset32),
    );
    registry.insert(
        Builder::new("Store")
            .imm(&imm.memflags)
            .value()
            .value()
            .imm(&imm.offset32),
    );
    registry.insert(
        Builder::new("StoreComplex")
            .imm(&imm.memflags)
            .value()
            .varargs()
            .imm(&imm.offset32),
    );
    registry.insert(Builder::new("StackLoad").imm(stack_slot).imm(&imm.offset32));
    registry.insert(
        Builder::new("StackStore")
            .value()
            .imm(stack_slot)
            .imm(&imm.offset32),
    );

    // Accessing a WebAssembly heap.
    registry.insert(Builder::new("HeapAddr").imm(heap).value().imm(&imm.uimm32));

    // Accessing a WebAssembly table.
    registry.insert(
        Builder::new("TableAddr")
            .imm(table)
            .value()
            .imm(&imm.offset32),
    );

    registry.insert(
        Builder::new("RegMove")
            .value()
            .imm_with_name("src", &imm.regunit)
            .imm_with_name("dst", &imm.regunit),
    );
    registry.insert(
        Builder::new("CopySpecial")
            .imm_with_name("src", &imm.regunit)
            .imm_with_name("dst", &imm.regunit),
    );
    registry.insert(Builder::new("CopyToSsa").imm_with_name("src", &imm.regunit));
    registry.insert(
        Builder::new("RegSpill")
            .value()
            .imm_with_name("src", &imm.regunit)
            .imm_with_name("dst", stack_slot),
    );
    registry.insert(
        Builder::new("RegFill")
            .value()
            .imm_with_name("src", stack_slot)
            .imm_with_name("dst", &imm.regunit),
    );

    registry.insert(Builder::new("Trap").imm(&imm.trapcode));
    registry.insert(Builder::new("CondTrap").value().imm(&imm.trapcode));
    registry.insert(
        Builder::new("IntCondTrap")
            .imm(&imm.intcc)
            .value()
            .imm(&imm.trapcode),
    );
    registry.insert(
        Builder::new("FloatCondTrap")
            .imm(&imm.floatcc)
            .value()
            .imm(&imm.trapcode),
    );

    registry
}
